pkg/syncer/metadatasyncer.go

```diff
@@ -551,6 +573,217 @@ func InitMetadataSyncer(ctx context.Context, clusterFlavor cnstypes.CnsClusterFl
+	if reflect.DeepEqual(oldTopoLabelsMap, newTopoLabelsMap) {
+		log.Debugf("topoCRUpdated: No change in %s CR topology labels. Ignoring the event",
+			csinodetopology.CRDSingular)
+		return
+	}
+	// Ideally a CSINodeTopology CR should never be updated after the status is set to Success but
+	// in cases where this does happen, in order to maintain the correctness of domainNodeMap, we
+	// will first remove the node name from previous topology labels before adding the new values.
+	if oldNodeTopoObj.Status.Status == csinodetopologyv1alpha1.CSINodeTopologySuccess {
+		log.Warnf("topoCRUpdated: %q instance with name %q has been updated after the Status was set to "+
+			"Success. Old object - %+v. New object - %+v", csinodetopology.CRDSingular, oldNodeTopoObj.Name,
+			oldNodeTopoObj, newNodeTopoObj)
+		removeLabelsFromTopologyVCMap(ctx, oldNodeTopoObj)
+	}
+	if newNodeTopoObj.Status.Status == csinodetopologyv1alpha1.CSINodeTopologySuccess {
+		addLabelsToTopologyVCMap(ctx, newNodeTopoObj)
+	}
+}
+
+// topoCRDeleted removes the topology to VC mapping for the deleted CSINodeTopology
+// instance from the MetadataSyncer.topologyVCMap.
+func topoCRDeleted(obj interface{}) {
+	ctx, log := logger.GetNewContextWithLogger()
+	// Verify object received.
+	var nodeTopoObj csinodetopologyv1alpha1.CSINodeTopology
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.(*unstructured.Unstructured).Object, &nodeTopoObj)
+	if err != nil {
+		log.Errorf("topoCRDeleted: failed to cast object %+v to %s type. Error: %+v",
+			csinodetopology.CRDSingular, err)
+		return
+	}
+	// Delete node name from domainNodeMap if the status of the CR was set to Success.
+	if nodeTopoObj.Status.Status == csinodetopologyv1alpha1.CSINodeTopologySuccess {
+		removeLabelsFromTopologyVCMap(ctx, nodeTopoObj)
+	} else {
+		log.Infof("topoCRDeleted: %q instance with name %q and status %q deleted.", csinodetopology.CRDSingular,
+			nodeTopoObj.Name, nodeTopoObj.Status.Status)
+	}
+}
+
+// removeLabelsFromTopologyVCMap removes the topology label to VC mapping for given CSINodeTopology
+// instance in the MetadataSyncer.topologyVCMap parameter.
+func removeLabelsFromTopologyVCMap(ctx context.Context, nodeTopoObj csinodetopologyv1alpha1.CSINodeTopology) {
+	log := logger.GetLogger(ctx)
+	nodeVM, err := nodeMgr.GetNode(ctx, nodeTopoObj.Spec.NodeUUID, nil)
+	if err != nil {
+		log.Errorf("Node %q is not yet registered in the node manager. Error: %+v",
+			nodeTopoObj.Spec.NodeUUID, err)
+		return
+	}
+	log.Infof("Removing VC %q mapping for TopologyLabels %+v.", nodeVM.VirtualCenterHost,
+		nodeTopoObj.Status.TopologyLabels)
+	for _, label := range nodeTopoObj.Status.TopologyLabels {
+		delete(MetadataSyncer.topologyVCMap[label.Value], nodeVM.VirtualCenterHost)
+	}
+}
+
+// getVCForTopologySegments uses the MetadataSyncer.topologyVCMap parameter to
+// retrieve the VC instance for the given topology segments map.
+// TODO: [Multi-VC] Remove the linter ignore once the function is used.
```

Comment:
Use it with unit test and remove this TODO

Reply:
Done